### PR TITLE
[SPARK-17159][STREAM] Significant speed up for running spark streaming against Object store.

### PR DESCRIPTION
**Pull request by ScrapCodes**
## What changes were proposed in this pull request?

Original work by Steve Loughran.
Based on #17745. 

This is a minimal patch of changes to FileInputDStream to reduce File status requests when querying files. Each call to file status is 3+ http calls to object store. This patch eliminates the need for it, by using FileStatus objects.

This is a minor optimisation when working with filesystems, but significant when working with object stores.

## How was this patch tested?

Tests included. Existing tests pass.

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
Hi, @ScrapCodes . Could you do the followings?
- Update the title to `[SPARK-17159][SS]...`
- Remove `Please review http://spark.apache.org/contributing.html ....` from PR description
- Share the numbers because the PR title has `Significant speed up`

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by ScrapCodes** [comment]
For numbers, while testing with object store having 50 files/dirs, without this patch it took 130 REST requests for 2 batches to complete and with this patch it took 56 rest requests. So number of rest calls are reduced, and this translates to speedup. How much speed up is dependent on number of files, but for the particular test, I have run, it was 2x.

**Comment by ScrapCodes** [comment]
Hi @srowen, would you like to take a look? Is there anything I can do, if this patch is missing something? I have tested it thoroughly against an object store.

File: streaming/src/main/scala/org/apache/spark/streaming/dstream/FileInputDStream.scala [hunk 1]
```diff
@@ -196,29 +191,29 @@ class FileInputDStream[K, V, F <: NewInputFormat[K, V]](
...
-        override def accept(path: Path): Boolean = fs.getFileStatus(path).isDirectory
-      }
-      val directories = fs.globStatus(directoryPath, directoryFilter).map(_.getPath)
+      val directories = Option(fs.globStatus(directoryPath)).getOrElse(Array.empty[FileStatus])
+          .filter(_.isDirectory)
+          .map(_.getPath)
       val newFiles = directories.flatMap(dir =>
-        fs.listStatus(dir, newFileFilter).map(_.getPath.toString))
+        fs.listStatus(dir)
+            .filter(isNewFile(_, currentTime, modTimeIgnoreThreshold))
```
Comment (srowen) [inline streaming/src/main/scala/org/apache/spark/streaming/dstream/FileInputDStream.scala]: Nit: I think the indent is too deep here?

File: streaming/src/main/scala/org/apache/spark/streaming/dstream/FileInputDStream.scala [hunk 2]
```diff
@@ -196,29 +191,29 @@ class FileInputDStream[K, V, F <: NewInputFormat[K, V]](
...
+            .map(_.getPath.toString))
       val timeTaken = clock.getTimeMillis() - lastNewFileFindingTime
-      logInfo("Finding new files took " + timeTaken + " ms")
-      logDebug("# cached file times = " + fileToModTime.size)
+      logInfo(s"Finding new files took $timeTaken ms")
```
Comment (srowen) [inline streaming/src/main/scala/org/apache/spark/streaming/dstream/FileInputDStream.scala]: I wonder if this should be a debug statement. I don't feel strongly about it.

File: streaming/src/main/scala/org/apache/spark/streaming/dstream/FileInputDStream.scala [hunk 3]
```diff
@@ -196,29 +191,29 @@ class FileInputDStream[K, V, F <: NewInputFormat[K, V]](
...
         s"ignoring files older than $modTimeIgnoreThreshold")
 
-      val newFileFilter = new PathFilter {
-        def accept(path: Path): Boolean = isNewFile(path, currentTime, modTimeIgnoreThreshold)
-      }
-      val directoryFilter = new PathFilter {
-        override def accept(path: Path): Boolean = fs.getFileStatus(path).isDirectory
-      }
-      val directories = fs.globStatus(directoryPath, directoryFilter).map(_.getPath)
+      val directories = Option(fs.globStatus(directoryPath)).getOrElse(Array.empty[FileStatus])
```
Comment (srowen) [inline streaming/src/main/scala/org/apache/spark/streaming/dstream/FileInputDStream.scala]: I guess the `.getOrElse` could come at the end, but it hardly matters.

**Review by srowen — APPROVED** [review]
If @steveloughran is into it, I think this is OK. I see why it's faster.

File: streaming/src/main/scala/org/apache/spark/streaming/dstream/FileInputDStream.scala (hunk 2 above)
Comment (steveloughran) [reply streaming/src/main/scala/org/apache/spark/streaming/dstream/FileInputDStream.scala]: depends on how big it grows over time

Comment (steveloughran) [reply streaming/src/main/scala/org/apache/spark/streaming/dstream/FileInputDStream.scala]: It was originally @ info, so if it it filled up logs *too much* there'd be complaints. What's important is that the time to scan is printed, either @ info or debug, so someone can see what's happening. Probably what does need logging @ warn is when the time to scan is greater than the window, or just getting close to it.

**Comment by srowen** [comment]
Yeah I agree, I was saying I do think it will speed things up. If it's a non-trivial win it's worthwhile even if it isn't the last optimization here. Is there any downside to this?

**Comment by steveloughran** [comment]
no, no cost penalties. Slightly lower namenode load too. If you had many, many spark streaming clients scanning directories, HDFS ops teams would eventually get upset. This will postpone the day

**Review by srowen — COMMENTED** [review]
@ScrapCodes looks good to me except perhaps the tiny style comment above, and possibly the log statement question

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Review by srowen — APPROVED** [review]

**Comment by ScrapCodes** [comment]
Thank you @srowen and @steveloughran.

**Pull request closed without merging**